File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/monitoring/config/MonitoringConfiguration.java
package com.study.webflux.rag.infrastructure.monitoring.config;

import java.util.List;

import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;

import com.study.webflux.rag.application.monitoring.monitor.LoggingPipelineMetricsReporter;
import com.study.webflux.rag.application.monitoring.monitor.PersistentPipelineMetricsReporter;
import com.study.webflux.rag.application.monitoring.monitor.PipelineMetricsReporter;
import com.study.webflux.rag.domain.monitoring.port.PerformanceMetricsRepository;
import com.study.webflux.rag.domain.monitoring.port.UsageAnalyticsRepository;
import com.study.webflux.rag.infrastructure.monitoring.micrometer.CompositePipelineMetricsReporter;
import com.study.webflux.rag.infrastructure.monitoring.micrometer.MicrometerPipelineMetricsReporter;

/** 파이프라인 모니터링 설정을 제공합니다. */
@Configuration
public class MonitoringConfiguration {

	/** 영구 저장 + Micrometer Reporter를 조합한 Composite Reporter를 생성합니다. */
	@Bean
	@Primary
	@ConditionalOnProperty(name = "monitoring.persistent.enabled", havingValue = "true", matchIfMissing = true)
	public PipelineMetricsReporter compositePipelineMetricsReporter(
		PerformanceMetricsRepository performanceMetricsRepository,
		UsageAnalyticsRepository usageAnalyticsRepository,
		LoggingPipelineMetricsReporter loggingReporter,
		MicrometerPipelineMetricsReporter micrometerReporter) {

		PersistentPipelineMetricsReporter persistentReporter = new PersistentPipelineMetricsReporter(
			performanceMetricsRepository,
			usageAnalyticsRepository,
			loggingReporter);

		return new CompositePipelineMetricsReporter(
			List.of(persistentReporter, micrometerReporter));
	}

	/** 영구 저장 비활성화 시 Micrometer Reporter만 사용합니다. */
	@Bean
	@ConditionalOnProperty(name = "monitoring.persistent.enabled", havingValue = "false")
	public PipelineMetricsReporter micrometerOnlyReporter(
		MicrometerPipelineMetricsReporter micrometerReporter,
		LoggingPipelineMetricsReporter loggingReporter) {
		return new CompositePipelineMetricsReporter(List.of(micrometerReporter, loggingReporter));
	}
}

File: infrastructure/src/main/kotlin/com/miyou/app/infrastructure/monitoring/config/MonitoringConfiguration.kt
package com.miyou.app.infrastructure.monitoring.config

import com.miyou.app.domain.monitoring.port.PerformanceMetricsRepository
import com.miyou.app.domain.monitoring.port.PipelineMetricsReporter
import com.miyou.app.domain.monitoring.port.UsageAnalyticsRepository
import com.miyou.app.infrastructure.monitoring.micrometer.CompositePipelineMetricsReporter
import com.miyou.app.infrastructure.monitoring.micrometer.MicrometerPipelineMetricsReporter
import com.miyou.app.infrastructure.outbound.monitoring.LoggingPipelineMetricsReporter
import com.miyou.app.infrastructure.outbound.monitoring.PersistentPipelineMetricsReporter
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary

/**
 * 모니터링 파이프라인 보고자 구성을 정의합니다.
 */
@Configuration
class MonitoringConfiguration {
    @Bean
    @Primary
    @ConditionalOnProperty(
        name = ["monitoring.persistent.enabled"],
        havingValue = "true",
        matchIfMissing = true,
    )
    fun compositePipelineMetricsReporter(
        performanceMetricsRepository: PerformanceMetricsRepository,
        usageAnalyticsRepository: UsageAnalyticsRepository,
        loggingReporter: LoggingPipelineMetricsReporter,
        micrometerReporter: MicrometerPipelineMetricsReporter,
    ): PipelineMetricsReporter {
        val persistentReporter =
            PersistentPipelineMetricsReporter(
                performanceMetricsRepository,
                usageAnalyticsRepository,
                loggingReporter,
            )
        return CompositePipelineMetricsReporter(
            listOf(
                persistentReporter,
                micrometerReporter,
            ),
        )
    }

    @Bean
    @ConditionalOnProperty(name = ["monitoring.persistent.enabled"], havingValue = "false")
    fun micrometerOnlyReporter(
        micrometerReporter: MicrometerPipelineMetricsReporter,
        loggingReporter: LoggingPipelineMetricsReporter,
    ): PipelineMetricsReporter =
        CompositePipelineMetricsReporter(
            listOf(
                micrometerReporter,
                loggingReporter,
            ),
        )
}
